Fuzzing update (#124)
ethereum/solidity-blog: _posts/2021-02-10-an-introduction-to-soliditys-fuzz-testing-approach.md

---
layout: post
published: true
title: 'An Introduction to Solidity's Fuzz Testing Approach'
date: '2021-02-10'
author: Bhargava Shastry
category: Explainers
---

Security vulnerabilities and bugs detract from software quality.
To discover them early, at best before they are released, we have adopted fuzz testing: feeding randomly generated programs to the Solidity compiler and observing the compilation runtime and code generated.

Since Q1 2019, the Solidity compiler is fuzz tested via Google's [open-source software fuzz][1] (oss-fuzz) framework.

In this post, we briefly describe the work that has been done on this front, and work that is currently in progress.

### Fuzzer Overview

Broadly speaking, we have developed two classes of fuzzers:
  - Front-end fuzzers that test the compiler front-end (parser/analyser).
  - Back-end fuzzers that test the compiler back-end (optimiser).

### Front-end Fuzzers

We have developed front-end fuzzers in order to test if programs are parsed and analysed correctly.
Front-end fuzzing comprises mutating existing test cases (most often unit tests) in order to test how parsing and program analysis respond to corner case input.

> The Solidity compiler makes heavy use of assertions during the input program parsing/analysis phase.

Reviewer: Actually mainly during code generation...

> The Solidity compiler makes heavy use of assertions during the input program parsing/analysis phase.
> Inputs generated by front-end fuzzers may fail such assertions, signalling an error in the compiler front-end.
> Please note that bugs that involve failed compiler assertions almost never lead to invalid code generation because code generation is initiated only when the parsing/analysis phase is successful.

Reviewer: ```suggestion
Fuzzing is effective because the Solidity compiler makes heavy use of assertions both during the input program parsing/analysis phase as well as the code generation phase.
Inputs generated by front-end fuzzers may fail such assertions, signalling an error in the compiler.
```
(Yeah, maybe we don't have to mention it that explicitly)
